csharp/src/Apache.Arrow/C/CArrowSchemaExporter.cs

```diff
@@ -0,0 +1,265 @@
+                case UInt16Type _: return "S";
+                case Int32Type _: return "i";
+                case UInt32Type _: return "I";
+                case Int64Type _: return "l";
+                case UInt64Type _: return "L";
+                // Floats
+                case HalfFloatType _: return "e";
+                case FloatType _: return "f";
+                case DoubleType _: return "g";
+                // Decimal
+                case Decimal128Type decimalType:
+                    return $"d:{decimalType.Precision},{decimalType.Scale}";
+                case Decimal256Type decimalType:
+                    return $"d:{decimalType.Precision},{decimalType.Scale},256";
+                // Binary
+                case BinaryType _: return "z";
+                case StringType _: return "u";
+                case FixedSizeBinaryType binaryType:
+                    return $"w:{binaryType.ByteWidth}";
+                // Date
+                case Date32Type _: return "tdD";
+                case Date64Type _: return "tdm";
+                // Time
+                case Time32Type timeType:
+                    return String.Format("tt{0}", FormatTimeUnit(timeType.Unit));
+                case Time64Type timeType:
+                    // Same prefix as Time32, but allowed time units are different.
+                    return String.Format("tt{0}", FormatTimeUnit(timeType.Unit));
+                // Timestamp
+                case TimestampType timestampType:
+                    return String.Format("ts{0}:{1}", FormatTimeUnit(timestampType.Unit), timestampType.Timezone);
+                // Nested
+                case ListType _: return "+l";
+                case StructType _: return "+s";
+                // Dictionary
+                case DictionaryType dictionaryType:
+                    return GetFormat(dictionaryType.IndexType);
+                default: throw new NotImplementedException($"Exporting {datatype.Name} not implemented");
+            };
+        }
+
+        private static long GetFlags(IArrowType datatype, bool nullable = true)
+        {
+            long flags = 0;
+
+            if (nullable)
+            {
+                flags |= CArrowSchema.ArrowFlagNullable;
+            }
+
+            if (datatype is DictionaryType dictionaryType)
+            {
+                if (dictionaryType.Ordered)
+                {
+                    flags |= CArrowSchema.ArrowFlagDictionaryOrdered;
+                }
+            }
+
+            // TODO: when we implement MapType, make sure to set the KEYS_SORTED flag.
+            return flags;
+        }
+
+        private static unsafe CArrowSchema** ConstructChildren(IArrowType datatype, out long numChildren)
+        {
+            if (datatype is NestedType nestedType)
+            {
+                IReadOnlyList<Field> fields = nestedType.Fields;
+                int numFields = fields.Count;
+                numChildren = numFields;
```

Comment:
Is there a chance that `numChildren` could be 0?  Does `AllocHGlobal` return `IntPtr.Zero` in that case?

Maybe more of a concern on import but I seem to recall a zero-field union being a useless, but technically possible, data type.